src/native/libs/System.Globalization.Native/pal_icushim_static.c

```diff
@@ -98,57 +115,332 @@ static int32_t load_icu_data(void* pData)
+        src_fd = -1;
+        icu_data_size = src_size;
+        icu_data_name = path;
+    }
+    const char *cache_mem = mmap(NULL, icu_data_size, PROT_READ, MAP_SHARED, cache_fd, 0);
+    if (cache_mem == MAP_FAILED)
+    {
+        log_shim_error("apple_mmap_icu_data: failed to map %s with size %zu", icu_data_name, icu_data_size);
+        goto error;
+    }
+    close(cache_fd);
+    close(dir_fd);
+    return cache_mem;
+error:
+    if (dir_fd >= 0)
+    {
+        close(dir_fd);
+    }
+    if (src_fd >= 0)
+    {
+        close(src_fd);
+    }
+    if (cache_fd >= 0)
+    {
+        close(cache_fd);
+    }
+    return NULL;
+}
+#else
+static const char *
+cstdlib_load_icu_data(const char *path)
+{
+    char *file_buf = NULL;
     FILE *fp = fopen(path, "rb");
-    if (fp == NULL) {
+
+    if (fp == NULL)
+    {
         log_shim_error("Unable to load ICU dat file '%s'.", path);
-        return ret;
+        goto error;
     }
 
-    if (fseek(fp, 0L, SEEK_END) != 0) {
-        fclose(fp);
+    if (fseek(fp, 0L, SEEK_END) != 0)
+    {
         log_shim_error("Unable to determine size of the dat file");
-        return ret;
+        goto error;
     }
 
-    long bufsize = ftell(fp);
+    long file_buf_size = ftell(fp);
 
-    if (bufsize == -1) {
-        fclose(fp);
+    if (file_buf_size == -1)
+    {
         log_shim_error("Unable to determine size of the ICU dat file.");
-        return ret;
+        goto error;
     }
 
-    icu_data = malloc(sizeof(char) * (bufsize + 1));
+    file_buf = malloc(sizeof(char) * (file_buf_size + 1));
 
-    if (icu_data == NULL) {
-        fclose(fp);
+    if (file_buf == NULL)
+    {
         log_shim_error("Unable to allocate enough to read the ICU dat file");
-        return ret;
+        goto error;
     }
 
-    if (fseek(fp, 0L, SEEK_SET) != 0) {
-        fclose(fp);
+    if (fseek(fp, 0L, SEEK_SET) != 0)
+    {
         log_shim_error("Unable to seek ICU dat file.");
-        return ret;
+        goto error;
     }
 
-    fread(icu_data, sizeof(char), bufsize, fp);
-    if (ferror( fp ) != 0 ) {
-        fclose(fp);
+    fread(file_buf, sizeof(char), file_buf_size, fp);
+    if (ferror( fp ) != 0)
+    {
         log_shim_error("Unable to read ICU dat file");
-        return ret;
+        goto error;
     }
 
     fclose(fp);
+    fp = NULL;
 
-    if (load_icu_data(icu_data) == 0) {
-        log_shim_error("ICU BAD EXIT %d.", ret);
-        return ret;
+    return file_buf;
+
+error:
+    if (fp != NULL)
+    {
+        fclose(fp);
+    }
+    if (file_buf != NULL)
+    {
+        free(file_buf);
+    }
+    return NULL;
+}
+#endif
+
+int32_t
+GlobalizationNative_LoadICUData(const char* path)
+{
+    const char *icu_data =
+        #if defined(USE_APPLE_DECOMPRESSION)
+        apple_mmap_icu_data(path)
+        #else
+        cstdlib_load_icu_data(path)
+        #endif
+        ;
+
+    if (icu_data == NULL)
+    {
+        log_shim_error("Failed to load ICU data.");
```

Comment:
Is this supposed to fall through and call `load_icu_data` even if there's no icu_data?

Reply:
No. This was an erroneous omission.